native-plasma/app/src/main/jni/renderer_rbga8888.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 2015 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#include "plasma_renderer.h"
+
+/* Color palette used for rendering the plasma */
+static const int PALETTE_BITS = 8;
+static const int PALETTE_SIZE = 1 << PALETTE_BITS;
+
+static uint32_t palette[PALETTE_SIZE];
+
+static uint32_t makeColor(int red, int green, int blue) {
+  return (uint32_t) (((0x00 << 24) & 0xFF000000) |
+                     ((blue << 16) & 0x00FF0000) |
+                     ((green << 8) & 0x0000FF00) |
+                     (red) & 0x000000FF);
+}
+
+void PlasmaRendererRGBA8888::init_palette() {
+  int nn, mm = 0;
+  // fun with colors
+  for (nn = 0; nn < PALETTE_SIZE / 4; nn++) {
+    int jj = (nn - mm) * 4 * 255 / PALETTE_SIZE;
+    palette[nn] = makeColor(255, jj, 255 - jj);
+  }
+
+  for (mm = nn; nn < PALETTE_SIZE / 2; nn++) {
+    int jj = (nn - mm) * 4 * 255 / PALETTE_SIZE;
+    palette[nn] = makeColor(255 - jj, 255, jj);
+  }
+
+  for (mm = nn; nn < PALETTE_SIZE * 3 / 4; nn++) {
+    int jj = (nn - mm) * 4 * 255 / PALETTE_SIZE;
+    palette[nn] = makeColor(0, 255 - jj, 255);
+  }
+
+  for (mm = nn; nn < PALETTE_SIZE; nn++) {
+    int jj = (nn - mm) * 4 * 255 / PALETTE_SIZE;
+    palette[nn] = makeColor(jj, 0, 255);
+  }
+}
+
+uint32_t PlasmaRendererRGBA8888::palette_from_fixed(Fixed x) {
```

Comment:
Well now that I think about it. The palette format is different for both
renderer, and the way you store the color is different. The only way to
achieve that would be to have a template in the parent class, and a
specialisation of template for the makeColor function (because the way to
store the color depend on the format of the data, uint16 or 32). It starts
to sound over engineered for a sample, knowing that most people doesn't
know well template and even less template specialisation. What do you think
?

On Fri, 11 Dec 2015 at 21:40 Johan Euphrosine <notifications@github.com>
wrote:

> In native-plasma/app/src/main/jni/renderer_rbga8888.cpp
> <https://github.com/googlesamples/android-ndk/pull/133#discussion_r47408594>
> :
>
> > +    int jj = (nn - mm) * 4 * 255 / PALETTE_SIZE;
> > +    palette[nn] = makeColor(255 - jj, 255, jj);
> > +  }
> > +
> > +  for (mm = nn; nn < PALETTE_SIZE * 3 / 4; nn++) {
> > +    int jj = (nn - mm) * 4 * 255 / PALETTE_SIZE;
> > +    palette[nn] = makeColor(0, 255 - jj, 255);
> > +  }
> > +
> > +  for (mm = nn; nn < PALETTE_SIZE; nn++) {
> > +    int jj = (nn - mm) * 4 * 255 / PALETTE_SIZE;
> > +    palette[nn] = makeColor(jj, 0, 255);
> > +  }
> > +}
> > +
> > +uint32_t PlasmaRendererRGBA8888::palette_from_fixed(Fixed x) {
>
> maybe this could also take parameter and be shared.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/googlesamples/android-ndk/pull/133/files#r47408594>.
>